src/commonpp/core/Gelf.cpp

```diff
@@ -0,0 +1,216 @@
+
+#include <boost/phoenix/bind.hpp>
+
+#include "commonpp/core/Utils.hpp"
+#include "commonpp/core/string/json_escape.hpp"
+#include "commonpp/thread/Thread.hpp"
+
+#include <boost/asio.hpp>
+#include <boost/asio/ip/udp.hpp>
+namespace logging = boost::log;
+namespace sinks = logging::sinks;
+namespace keywords = logging::keywords;
+namespace expr = logging::expressions;
+namespace attrs = logging::attributes;
+namespace src = logging::sources;
+
+namespace
+{
+
+size_t to_syslog_level(commonpp::LoggingLevel level)
+{
+    return 7 - static_cast<size_t>(level);
+}
+}
+
+namespace commonpp
+{
+namespace core
+{
+
+class GelfFormatter
+{
+    std::string host_;
+    std::vector<std::pair<std::string, std::string>> static_fields_;
+
+public:
+    GelfFormatter(std::string host,
+                  std::vector<std::pair<std::string, std::string>> static_fields)
+    : host_(std::move(host))
+    , static_fields_(std::move(static_fields))
+    {
+        // it might be possible to "pre-render" these to a fragment of json.
+        for (auto& s : static_fields_)
+        {
+            s.second = commonpp::string::escape_json_string(s.second);
+        }
+    }
+
+    void operator()(const logging::record_view& rec,
+                    logging::formatting_ostream& strm)
+    {
+        strm << "{";
+        strm << "\"version\" : \"1.1\",";
+        strm << "\"host\" : \"" << host_ << "\",";
+        for (auto s : static_fields_)
+        {
+            strm << "\"" << s.first << "\" : \"" << s.second << "\",";
+        }
+        strm << "\"short_message\" : \""
+             << commonpp::string::escape_json_string(*rec[expr::smessage])
+             << "\"";
+        strm << "}";
+    }
+};
+
+class GelfUDPBackend
+    : public sinks::basic_formatted_sink_backend<char, sinks::concurrent_feeding>
+{
+    boost::asio::io_service io_service_;
+    boost::asio::ip::udp::socket socket_;
+    std::string host_;
+    uint16_t port_;
+
+    static constexpr size_t MAX_PACKET_SIZE = 1000;
+    static constexpr size_t HEADER_OVERHEAD = 12;
+
+public:
+    using base =
+        sinks::basic_formatted_sink_backend<char, sinks::concurrent_feeding>;
+
+    GelfUDPBackend(std::string host, uint16_t port)
+    : base()
+    , socket_(io_service_)
+    , host_(std::move(host))
+    , port_(port)
+    {
+        try
+        {
+            using udp = boost::asio::ip::udp;
+            udp::resolver resolver(io_service_);
+            udp::resolver::query query(host_, "");
+
+            udp::resolver::iterator r = resolver.resolve(query);
+            udp::resolver::iterator end;
+
+            if (r == end)
+            {
+                throw std::runtime_error("Cannot resolve GELF host address: " +
+                                         host_);
+            }
+
+            udp::endpoint endpoint = *r;
+            endpoint.port(port);
+            socket_.connect(endpoint);
+        }
+        catch (std::exception& e)
+        {
+            throw;
```

Comment:
why catch and throw again ?

Reply:
lol. I think at one time I had a logging statement in there but took it out.